Userland/Libraries/LibAudio/PlaybackStreamOboe.cpp

```diff
@@ -0,0 +1,123 @@
+    m_last_sample_time.store(last_sample_time);
+    Bytes output_buffer {
+        reinterpret_cast<u8*>(audioData),
+        static_cast<size_t>(numFrames * oboeStream->getChannelCount() * sizeof(float))
+    };
+    auto written_bytes = m_data_request_callback(output_buffer, PcmSampleFormat::Float32, numFrames);
+    if (written_bytes.is_empty())
+        return oboe::DataCallbackResult::Stop;
+
+    m_number_of_samples_enqueued += numFrames;
+    float* output = (float*)audioData;
+    for (int frames = 0; frames < numFrames; frames++) {
+        for (int channels = 0; channels < oboeStream->getChannelCount(); channels++) {
+            *output++ *= m_volume.load();
+        }
+    }
+    return oboe::DataCallbackResult::Continue;
+}
+
+void OboeCallback::set_volume(float volume)
+{
+    m_volume.store(volume);
+}
+
+OboeCallback::OboeCallback(PlaybackStream::AudioDataRequestCallback data_request_callback)
+    : m_data_request_callback(move(data_request_callback))
+{
+}
+
+PlaybackStreamOboe::PlaybackStreamOboe(std::shared_ptr<oboe::AudioStream> stream, std::shared_ptr<OboeCallback> oboe_callback)
+    : m_stream(move(stream))
+    , m_oboe_callback(move(oboe_callback))
+{
+}
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamOboe::create(OutputState initial_output_state, u32 sample_rate, u8 channels, u32, AudioDataRequestCallback&& data_request_callback)
+{
+    std::shared_ptr<oboe::AudioStream> stream;
+    auto oboe_callback = std::make_shared<OboeCallback>(move(data_request_callback));
+    oboe::AudioStreamBuilder builder;
+    auto result = builder.setSharingMode(oboe::SharingMode::Shared)
+                      ->setPerformanceMode(oboe::PerformanceMode::LowLatency)
+                      ->setFormat(oboe::AudioFormat::Float)
+                      ->setDataCallback(oboe_callback)
+                      ->setChannelCount(channels)
+                      ->setSampleRate(sample_rate)
+                      ->openStream(stream);
+
+    if (result != oboe::Result::OK)
+        return Error::from_string_literal("Oboe failed to start");
+
+    if (initial_output_state == OutputState::Playing)
+        stream->requestStart();
+
+    return TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamOboe(move(stream), move(oboe_callback))));
+}
+
+PlaybackStreamOboe::~PlaybackStreamOboe() = default;
+
+void PlaybackStreamOboe::set_underrun_callback(Function<void()>)
+{
+    // FIXME: Implement this.
+}
+
+NonnullRefPtr<Core::ThreadedPromise<Duration>> PlaybackStreamOboe::resume()
+{
+    auto promise = Core::ThreadedPromise<Duration>::create();
+    auto time = MUST(total_time_played());
+    m_stream->requestStart();
+    promise->resolve(move(time));
+    return promise;
+}
+
+NonnullRefPtr<Core::ThreadedPromise<void>> PlaybackStreamOboe::drain_buffer_and_suspend()
+{
+    auto promise = Core::ThreadedPromise<void>::create();
+    m_stream->requestPause();
+    promise->resolve();
```

Comment:
The request is asynchronously made. It seems correct since the Serenity backend does like it.

Reply:
The Serenity backend is extremely primitive due to limitations in the code generated for the IPC interface for the AudioServer. It would be best to reference what the PulseAudio interface does instead, or just meet the requirements that the comments in the PlaybackStream.h file prescribe (if I wrote those to the necessary specificity).

Reply:
Ok, I made the functions synchronous.